src/Microsoft.AspNet.Razor/SourceLocation.cs

```diff
@@ -89,57 +166,79 @@ public static SourceLocation Add(SourceLocation left, SourceLocation right)
             else
             {
                 return new SourceLocation(
+                    left.FilePath,
                     left.AbsoluteIndex + right.AbsoluteIndex,
                     left.LineIndex + right.LineIndex,
                     left.CharacterIndex + right.CharacterIndex);
             }
         }
 
-        public static SourceLocation Subtract(SourceLocation left, SourceLocation right)
+        /// <summary>
+        /// Subtracts two <see cref="SourceLocation"/>s.
+        /// </summary>
+        /// <param name="left">The left operand.</param>
+        /// <param name="right">The right operand.</param>
+        /// <returns>A <see cref="SourceLocation"/> that is the difference of the left and right operands.</returns>
+        /// <exception cref="ArgumentException">if the <see cref="FilePath"/> of the left and right operands
+        /// are different.</exception>
+        public static SourceLocation operator -(SourceLocation left, SourceLocation right)
         {
+            if (!string.Equals(left.FilePath, right.FilePath, StringComparison.Ordinal))
+            {
+                throw new ArgumentException(
+                    RazorResources.FormatSourceLocationFilePathDoesNotMatch(nameof(SourceLocation)),
+                    nameof(right));
+            }
+
             return new SourceLocation(
+                left.FilePath,
```

Comment:
including the file in a relative location seems a bit nonsensical.  one possibility would be for `-` to return a `SourceLocation` with `FilePath==null` (maintaining current comparison of input `FilePath` values) and `+` to `throw` if _both_ `SourceLocations` have `FilePath!=null`, then return a new `SourceLocation` with the non-`null` `FilePath` (if any).